### ISSUE TITLE
A "Rendered term" format display for term reference fields

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#3283]
## Describe your issue or idea

It would be useful to have a "Rendered term" field format for the display of term reference fields. Similar to how it already exists for node references (and entity references). Example from node reference (the "Rendered Node" format for "Webform" field:

![screenshot from 2018-09-09 22-45-05](https://user-images.githubusercontent.com/1071483/45273783-2f78e380-b482-11e8-8f88-9633d12993f5.png)

This feature in node references provides an easy way to display the referenced content rather than just a link or title. There was a recent post in the forum hoping to do this but it's not currently possible in Backdrop. Actually it can be done with Views but it's not super obvious or easy.

----
PR by @herbdool: https://github.com/backdrop/backdrop/pull/2301
Contrib module: https://github.com/backdrop-contrib/taxonomy_field_formatter

**Comment by herbdool** [comment]
This was not too hard. Adapted code from node references and it works well for taxonomy terms. Since it just provides a new display option it won't affect any existing functionality.

![screenshot from 2018-09-09 22-58-17](https://user-images.githubusercontent.com/1071483/45274174-2be65c00-b484-11e8-95e9-1411a2bf5886.png)

**Comment by herbdool** [comment]
@opi want to review again?

**Comment by opi** [comment]
Reviewed and tested, works well !

I'm not sure I can review the `$recursion_queue` part ([link](https://github.com/backdrop/backdrop/blob/d0c3ced46b453374e2c634ecf216764f73dd1773/core/modules/taxonomy/taxonomy.module#L1493-L1553)). The `taxonomy_field_formatter` contrib module does not have one, and works good too. It may come from the `references` module, and should be needed for recursion edge case.

**Comment by herbdool** [comment]
Thanks @opi! Yes, the recursion code comes from References module. Can hopefully help where a term has a term field. We don't want to render them ad infinitum.

**Comment by opi** [comment]
@herbdool should we get this issue milestoned (or mentionned in the next dev meeting) to get some traction, and not to be forgotten ?

**Comment by herbdool** [comment]
Good idea. I can add milestones. Can you add it to the dev agenda?

**Comment by quicksketch** [comment]
This looks great. FYI being marked RTBC certainly makes it get attention. :smile:

I left two comments at https://github.com/backdrop/backdrop/pull/2301#pullrequestreview-159599948.

The recursion part also gives me a little trouble. And I'm not sure it would work with core's term reference fields because they don't set a `referencing_entity` property.

Perhaps instead of checking that property, we could set a recursion "count" variable, and only allow the nesting of a referenced term two-levels deep (probably with a variable to increase if needed). Something like this:

```
$recursion_count = &backdrop_static(__FUNCTION__, 0);
$recursion_count++;

if ($recursion_count > 2) {
  // Use only term title.
}
else {
  // Fully rendered term.
}

$recursion_count--;
```

This does mean that a term with a term reference field would only show two levels deep before stopping, even if there wasn't a recursive loop, but it would work with current taxonomy term reference fields.

**Comment by herbdool** [comment]
@quicksketch I've checked and it does seem like the recursion works. I don't really understand the comment about the "bug" in the code either (copied from node_reference) so I removed the comment. It seems like it's more like a limitation rather than a bug. Only reference I could find is this complicated to recreate issue https://www.drupal.org/project/references/issues/1485328.

On a vocabulary I have a category field that uses it's own vocabulary. If I reference a term which references the same parent term it will show the second term reference as a link:

![screenshot from 2018-09-29 15-38-49](https://user-images.githubusercontent.com/1071483/46249880-98f36e80-c3fe-11e8-8867-8e60489ae9ae.png)

But if the second term reference is a different term from the parent then it's fine:

![screenshot from 2018-09-29 15-38-17](https://user-images.githubusercontent.com/1071483/46249883-b0caf280-c3fe-11e8-9b2e-7d404dba716f.png)

So "referencing_entity" seems to be working fine as a temporary property of the object.

I updated the PR.
